fix(router): Ensures guard references are cleared after navigation

SkyZeroZx opened the pull request:
Allows component instances to be garbage collected, improving resource use in long-running apps

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number:  https://github.com/angular/angular/issues/63983


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Previous behavior
Old references were retained after navigation.
<img width="1429" height="271" alt="image" src="https://github.com/user-attachments/assets/fcd31e56-7a8d-41cf-8982-1dbc06cfac9e" />




Old references are now properly cleaned up when navigating, 
<img width="1427" height="358" alt="image" src="https://github.com/user-attachments/assets/4330f282-540c-4a64-ace5-f95ece0f52fa" />

atscott commented:
Okay, then the controller needs to be pulled out of the transition and created in the same or similar place to completedOrAborted. None of the internal transition should be retained

SkyZeroZx commented:
I tried replacing the abort reference and tested in Chrome, Edge, and Brave, but the old component reference was still not being cleaned up.
Could it be that the object reference is being retained somewhere else?
So far, the most reliable solution I’ve found is to clear it inside finalize.

```ts
      switchMap((overallTransitionState) => {
        let completedOrAborted = false;
        // Create an AbortController local to this transition processing scope
        // so closures (like the public `abort` function) can capture only
        // the controller reference and not the whole transition object.
        let abortControllerRef = new AbortController();
        overallTransitionState.abortController = abortControllerRef;
```

atscott commented:
Did you also remove the reference to the transition in the abort method and refer directly to abortControllerRef? I don’t see how it could retain a reference to the object if it’s never referenced at all

SkyZeroZx commented:
I was wondering the same. If I update all the corresponding references so that AbortController is used directly at the completedOrAborted level, then the navigation object only stores a reference to abortControllerRef.abort()

```ts
  const lastSuccessfulNavigation = this.lastSuccessfulNavigation();
            // Store the Navigation object
            this.currentNavigation.set({
              id: t.id,
              initialUrl: t.rawUrl,
              extractedUrl: t.extractedUrl,
              targetBrowserUrl:
                typeof t.extras.browserUrl === 'string'
                  ? this.urlSerializer.parse(t.extras.browserUrl)
                  : t.extras.browserUrl,
              trigger: t.source,
              extras: t.extras,
              previousNavigation: !lastSuccessfulNavigation
                ? null
                : {
                    ...lastSuccessfulNavigation,
                    previousNavigation: null,
                  },
              abort: () => abortControllerRef.abort(),
            });
 ```
 
 ```ts
   // Recognize
                recognize(
                  this.environmentInjector,
                  this.configLoader,
                  this.rootComponentType,
                  router.config,
                  this.urlSerializer,
                  this.paramsInheritanceStrategy,
                  abortControllerRef.signal,
                ),
```
```ts 
takeUntil(
            abortSignalToObservable(abortControllerRef.signal).pipe(
              // Ignore aborts if we are already completed, canceled, or are in the activation stage (we have targetRouterState)
              filter(() => !completedOrAborted && !overallTransitionState.targetRouterState),
              tap(() => {
                this.cancelNavigationTransition(
                  overallTransitionState,
                  abortControllerRef.signal.reason + '',
                  NavigationCancellationCode.Aborted,
                );
              }),
            ),
          ),
```

```ts
  finalize(() => {
           abortControllerRef.abort();
            /* When the navigation stream finishes either through error or success,
             * we set the `completed` or `errored` flag. However, there are some
             * situations where we could get here without either of those being set.
             * For instance, a redirect during NavigationStart. Therefore, this is a
             * catch-all to make sure the NavigationCancel event is fired when a
             * navigation gets cancelled but not caught by other means. */
            if (!completedOrAborted) {
              const cancelationReason =
                typeof ngDevMode === 'undefined' || ngDevMode
                  ? `Navigation ID ${overallTransitionState.id} is not equal to the current navigation id ${this.navigationId}`
                  : '';
              this.cancelNavigationTransition(
                overallTransitionState,
                cancelationReason,
                NavigationCancellationCode.SupersededByNewNavigation,
              );
            }
            // Only clear current navigation if it is still set to the one that
            // finalized.
            if (this.currentTransition?.id === overallTransitionState.id) {
              this.currentNavigation.set(null);
              this.currentTransition = null;
            }
          }),  
```

SkyZeroZx commented:
I’m not sure if I explained myself clearly. Even after applying the changes mentioned above, the references to components from previous routes are still being retained.

In my tests, the only reliable way to remove these references was by clearing them inside finalize.
```ts
finalize(() => {
            overallTransitionState.abortController.abort();
            /* When the navigation stream finishes either through error or success,
             * we set the `completed` or `errored` flag. However, there are some
             * situations where we could get here without either of those being set.
             * For instance, a redirect during NavigationStart. Therefore, this is a
             * catch-all to make sure the NavigationCancel event is fired when a
             * navigation gets cancelled but not caught by other means. */
            if (!completedOrAborted) {
              const cancelationReason =
                typeof ngDevMode === 'undefined' || ngDevMode
                  ? `Navigation ID ${overallTransitionState.id} is not equal to the current navigation id ${this.navigationId}`
                  : '';
              this.cancelNavigationTransition(
                overallTransitionState,
                cancelationReason,
                NavigationCancellationCode.SupersededByNewNavigation,
              );
            }
            // Only clear current navigation if it is still set to the one that
            // finalized.
            if (this.currentTransition?.id === overallTransitionState.id) {
              this.currentNavigation.set(null);
              this.currentTransition = null;
            }
            // Clear the reference to the AbortController
            (overallTransitionState as any) = null;
          })
          
  ```

SkyZeroZx committed:
fix(router): Ensures guard references are cleared after navigation

Allows component instances to be garbage collected, improving resource use in long-running apps
